weather/weather.go

```diff
@@ -12,56 +10,51 @@ const (
 	weatherURL  = "https://api.openweathermap.org/data/2.5/weather?appid="
 )
 
-type Client struct {
+type WeatherClient struct {
 	apiKey string
 	unit   string
 
-	common service
-
-	Temperature *TemperatureService
-	Location    *LocationService
+	httpClient *resty.Client
 }
 
-type service struct {
-	client *Client
+type WeatherConfig struct {
+	ApiKey string
+	Units  string "metric" // This isn't setting a default
 }
 
-func NewClient(apiKey string, unit string) *Client {
-	c := &Client{
-		apiKey: apiKey,
-		unit:   unit,
-	}
-	if c.unit == "" {
-		c.unit = "metric"
+func NewClient(config WeatherConfig) (*WeatherClient, error) {
+
+	if config.ApiKey == "" {
+		cause := errors.New(apiKeyNotSupplied)
+		return nil, errors.WithStack(cause)
 	}
-	c.initialize()
-	return c
-}
 
-func (c *Client) initialize() {
-	c.common.client = c
+	if config.Units == "" {
+		config.Units = "metric"
+	}
 
-	c.Temperature = (*TemperatureService)(&c.common)
-	c.Location = (*LocationService)(&c.common)
+	c := &WeatherClient{
+		apiKey: config.ApiKey,
+		unit:   config.Units,
+	}
+	c.httpClient = resty.New()
+	return c, nil
 }
 
-// TODO Switch to using resty
-func (c *Client) NewRequest(url string) ([]byte, error) {
+func (c *WeatherClient) Fetch(url string) ([]byte, error) {
+
+	resp, err := c.httpClient.R().
+		EnableTrace().
+		Get(url)
 
-	resp, err := http.Get(url)
 	if err != nil {
 		return nil, err
 	}
 
-	if resp.StatusCode == 401 {
+	if resp.StatusCode() == 401 {
 		cause := errors.New(invalidAPIKey)
 		return nil, errors.WithStack(cause)
```

Comment:
`errors.New` adds a stack trace and `cause` is never used again so you could replace both these lines with a single line that does `return nil, errors.New(invalidAPIKey)`